src/backend/distributed/shardsplit/cdc_decoder.c

```diff
@@ -0,0 +1,388 @@
+								  shardId, Oid targetRelationid)
+{
+	/* Get the distributed table's relation for this shard.*/
+	Relation targetRelation = RelationIdGetRelation(targetRelationid);
+
+	/*
+	 * Check if there has been a schema change (such as a dropped column), by comparing
+	 * the number of attributes in the shard table and the shell table.
+	 */
+	TranslateChangesIfSchemaChanged(relation, targetRelation, change);
+
+	/*
+	 * Publish the change to the shard table as the change in the distributed table,
+	 * so that the CDC client can see the change in the distributed table,
+	 * instead of the shard table, by calling the pgoutput's callback function.
+	 */
+	ouputPluginChangeCB(ctx, txn, targetRelation, change);
+	RelationClose(targetRelation);
+}
+
+
+/*
+ * PublishChangesIfCdcSlot checks if the current slot is a CDC slot. If so, it publishes
+ * the changes as the change for the distributed table instead of shard.
+ * If not, it returns false. It also skips the Citus metadata tables.
+ */
+void
+PublishDistributedTableChanges(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
+							   Relation relation, ReorderBufferChange *change)
+{
+	char *shardRelationName = RelationGetRelationName(relation);
+
+	/* Skip publishing CDC changes for any system relations in pg_catalog*/
+	if (relation->rd_rel->relnamespace == PG_CATALOG_NAMESPACE)
+	{
+		return;
+	}
+
+	/* Check if the relation is a distributed table by checking for shard name.	*/
+	uint64 shardId = ExtractShardIdFromTableName(shardRelationName, true);
+
+	/* If this relation is not distributed, call the pgoutput's callback and return. */
+	if (shardId == INVALID_SHARD_ID)
+	{
+		ouputPluginChangeCB(ctx, txn, relation, change);
+		return;
+	}
+
+	bool isReferenceTable = false;
+	Oid distRelationId = LookupDistributedTableIdForShardId(shardId, &isReferenceTable);
+	if (distRelationId == InvalidOid)
+	{
+		ouputPluginChangeCB(ctx, txn, relation, change);
+		return;
+	}
+
+	/* Publish changes for reference table only from the coordinator node. */
+	if (isReferenceTable && !IsCoordinator())
+	{
+		return;
+	}
+
+	/* translate and publish from shard relation to distributed table relation for CDC. */
+	TranslateAndPublishRelationForCDC(ctx, txn, relation, change, shardId,
+									  distRelationId);
+}
+
+
+/*
+ * GetTupleForTargetSchemaForCdc returns a heap tuple with the data from sourceRelationTuple
+ * to match the schema in targetRelDesc. Either or both source and target relations may have
+ * dropped columns. This function handles it by adding NULL values for dropped columns in
+ * target relation and skipping dropped columns in source relation. It returns a heap tuple
+ * adjusted to the current schema of the target relation.
+ */
+static HeapTuple
+GetTupleForTargetSchemaForCdc(HeapTuple sourceRelationTuple,
+							  TupleDesc sourceRelDesc,
+							  TupleDesc targetRelDesc)
+{
+	/* Allocate memory for sourceValues and sourceNulls arrays. */
```

Comment:
an optimization we could make here is that if neither side has dropped columns, we don't need to do anything

Reply:
nvm, already there :)